### PR TITLE
japanese plugin v1.0

### PR DESCRIPTION
**Pull request by YS-jack**
Can translates text of most if not all widgets into japanese, can also chat in japanese. can use DeepL api to translate thing more accurately.
This plugin is targeted for Japanese players who can't read English but want to play this game.

Currently some widgets are set not to be translated because I havent provided a translation for that part, and bad translation can cause confusion.

Can selectively choose not to translate if desired in config section.

### CODE REVIEW

**Comment by YS-jack** [comment]
could someone help me with this error?
i have added third party dependency then ran 
`gradlew --write-verification-metadata sha256`
after adding deepL dependency in [package/verification-template/build.gradle](https://github.com/runelite/plugin-hub/blob/master/package/verification-template/build.gradle)

`FAILURE: Build failed with an exception.

* Where:
Initialization script '/home/runner/work/plugin-hub/plugin-hub/package/target_init.gradle' line: 61

* What went wrong:
Could not determine the dependencies of task ':runelitePluginHubPackage'.
> Dependency verification failed for configuration ':runtimeClasspath'
  2 artifacts failed verification:
    - annotations-20.1.0.jar (org.jetbrains:annotations:20.1.0) from repository MavenRepo
    - annotations-20.1.0.pom (org.jetbrains:annotations:20.1.0) from repository MavenRepo
  If the artifacts are trustworthy, you will need to update the gradle/verification-metadata.xml file by following the instructions at https://docs.gradle.org/7.4/userguide/dependency_verification.html#sec:troubleshooting-verification
  
  Open this report for more details: file:///tmp/pluginhub-package6013681569718307465/japanese/repo/build/reports/dependency-verification/at-1712326246275/dependency-verification-report.html

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
build: 6498ms
package failed
net.runelite.pluginhub.packager.PluginBuildException: japanese: build failed
	at net.runelite.pluginhub.packager.PluginBuildException.of(PluginBuildException.java:54)
	at net.runelite.pluginhub.packager.Plugin.build(Plugin.java:536)
	at net.runelite.pluginhub.packager.Packager.buildPlugin(Packager.java:215)
	at net.runelite.pluginhub.packager.Packager.lambda$buildPlugins$0(Packager.java:[155](https://github.com/runelite/plugin-hub/actions/runs/8571115412/job/23490638822?pr=5809#step:6:156))
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.gradle.tooling.BuildException: Could not execute build using connection to Gradle installation '/home/runner/.gradle/wrapper/dists/gradle-7.4-all/aadb4xli5jkdsnukm30eibyiu/gradle-7.4'.
	...
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	... 1 more
Caused by: org.gradle.internal.exceptions.LocationAwareException: Initialization script '/home/runner/work/plugin-hub/plugin-hub/package/target_init.gradle' line: 61
Could not determine the dependencies of task ':runelitePluginHubPackage'.
	...
Caused by: org.gradle.api.internal.tasks.TaskDependencyResolveException: Could not determine the dependencies of task ':runelitePluginHubPackage'.
	...
	... 75 more
Caused by: org.gradle.api.internal.artifacts.verification.DependencyVerificationException: Dependency verification failed for configuration ':runtimeClasspath'
2 artifacts failed verification:
  - annotations-20.1.0.jar (org.jetbrains:annotations:20.1.0) from repository MavenRepo
  - annotations-20.1.0.pom (org.jetbrains:annotations:20.1.0) from repository MavenRepo
If the artifacts are trustworthy, you will need to update the gradle/verification-metadata.xml file by following the instructions at https://docs.gradle.org/7.4/userguide/dependency_verification.html#sec:troubleshooting-verification

Open this report for more details: file:///tmp/pluginhub-package601368[156](https://github.com/runelite/plugin-hub/actions/runs/8571115412/job/23490638822?pr=5809#step:6:157)9718307465/japanese/repo/build/reports/dependency-verification/at-1712326246275/dependency-verification-report.html
	...
	at target_init_6d8nx9ehfn6xp8roxihe52luu$_run_closure1$_closure6$_closure10.doCall(/home/runner/work/plugin-hub/plugin-hub/package/target_init.gradle:61)
	at target_init_6d8nx9ehfn6xp8roxihe52luu$_run_closure1$_closure6$_closure10.doCall(/home/runner/work/plugin-hub/plugin-hub/package/target_init.gradle)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	...
	... 143 more`

**Comment by LlemonDuck** [comment]
remove the gradle wrappers you committed

**Comment by YS-jack** [comment]
> remove the gradle wrappers you committed

is that the one in plugin-hub, or in my plugin?

**Comment by LlemonDuck** [comment]
Keep this line `thirdParty("com.deepl.api:deepl-java:1.4.0") { because "japanese" }`

Delete 
* `package/verification-template/gradle/wrapper/gradle-wrapper.jar`
* `package/verification-template/gradle/wrapper/gradle-wrapper.properties`
* `package/verification-template/gradlew`
* `package/verification-template/gradlew.bat`

**Comment by Nightfirecat** [comment]
Is this something which could be added to the existing [Polywoof hub plugin](https://runelite.net/plugin-hub/show/polywoof)? It already supports translation via DeepL as one of its backends is why I suggest it.

**Comment by YS-jack** [comment]
This plugin does many different things from the polywoof's translation plugin, and there are also reasons it cant be added to polywoof:
- this plugin translates the widget text and menu entries, but polywoof only creates an overlay which displays translation. Looking at polywoof  README, it probably won't accept my plugin's feature to be added:
> [Widget Text and UI are untouched to prevent confusion of misleading translations](https://github.com/furfy/polywoof/tree/master?tab=readme-ov-file#user-content-fn-4-caf2e44e4c03ff0a9d6c39f99b6f5971)
- the file size is very close to the 10Mib limit because the Japanese characters are inserted as png images. 
  - The file size is huge because there are many Japanese characters, multiplied by the number of colors. Originally wanted to include 8 colours, but because of the limit I reduced it to 4, also discarded many uncommon characters which made up 30% of the whole set.
- plugin's settings are in Japanese because it assumes the player doesnt konw any English. 
  - I want to make OSRS accesible to Japanese gamers because I'm sure they'll love it, but I dare say many of them are allergic to English. 
  - You might ask if theres a demand for such Japanese translation project, but I have seen few reddit posts asking for Japanese translations, and also know people who want to invite their Japanese friends or use it themselves on X and my Japanese RS discord server

**Comment by raiyni** [comment]
I don't think this plugin will work how you think it does as you aren't using resource directories when accessing your translate data

**Comment by YS-jack** [comment]
> I don't think this plugin will work how you think it does as you aren't using resource directories when accessing your translate data

do you mean I must place all resources (images, translations) directly under src/main/resources directory, and under src/main/resources/someFolder/ ?

**Comment by Nightfirecat** [comment]
No, the issue is you're accessing the images via the filesystem, rather than from the resources, see: https://github.com/runelite/plugin-hub?tab=readme-ov-file#plugin-resources

Basically, this only works because those png files exist on your filesystem. If you were to build this plugin into a jar and run it on another computer, or remove your copy of the repository from your filesystem, this would encounter errors trying to read the now-nonexistent files.

**Comment by YS-jack** [comment]
thanks, I made the difference for reading files. Im guessing i can leave the code for writing to file

File: package/verification-template/build.gradle [hunk 1]
```diff
@@ -18,7 +18,7 @@ configurations {
 dependencies {
 	core group: "net.runelite", name: "client", version: file("../../runelite.version").text.trim()
 	core "org.projectlombok:lombok:1.18.30"
-	core "org.jetbrains:annotations:23.0.0"
+	core "org.jetbrains:annotations:20.0.0"
```
Comment (Nightfirecat) [inline package/verification-template/build.gradle:21]: What is this downgraded for?

**Comment by Nightfirecat** [comment]
The file writing should be changed to a path under `~/.runelite`. You can lean on the `RUNELITE_DIR` constant from the client, eg. `new File(RUNELITE_DIR, "japanese")`, for a location to write to. As-is, it looks like it would write to some location on the user's system relative to the cwd where RuneLite was run from instead of a fixed path.
[Don't use `enabledByDefault = false`](https://github.com/YS-jack/japanesePlugin/blob/457d2479602cb6c971d9248ae94f9eee952ef639/src/main/java/com/japanese/JapanesePlugin.java#L33) as it's confusing for users to install a plugin which isn't enabled right away.

I haven't yet reviewed the code of the deepl dependency, but the rest of this pulgin's code seems fine to me.

File: package/verification-template/build.gradle (hunk 1 above)
Comment (YS-jack) [reply package/verification-template/build.gradle]: first few attempts to send pull request failed, so I was comparing my code with other's, and found that this version on mine was different so I downgraded it to someone elses

I changed it back to 23.0.0, and it seems fine now

**Comment by YS-jack** [comment]
will work on the other things you mentioned from tomorrow, but i have undone the version changes

> The file writing should be changed to a path under `~/.runelite`. You can lean on the `RUNELITE_DIR` constant from the client, eg. `new File(RUNELITE_DIR, "japanese")`, for a location to write to. As-is, it looks like it would write to some location on the user's system relative to the cwd where RuneLite was run from instead of a fixed path. [Don't use `enabledByDefault = false`](https://github.com/YS-jack/japanesePlugin/blob/457d2479602cb6c971d9248ae94f9eee952ef639/src/main/java/com/japanese/JapanesePlugin.java#L33) as it's confusing for users to install a plugin which isn't enabled right away.
> 
> I haven't yet reviewed the code of the deepl dependency, but the rest of this pulgin's code seems fine to me.

**Comment by YS-jack** [comment]
well, that didnt work..
ill try resolve it from "resolve conflicts" button on

**Comment by YS-jack** [comment]
changes I made to my plugin itself:
all files that were in the resource folder are now instead downloaded from a [github repository I created](https://github.com/YS-jack/JapanesePluginData/tree/main/repos) to the external file (inside RUNELITE_DIR)
So they are read and wrote not as stream but as normal files.

because of this, I do
`final BufferedImage image = ImageIO.read(externalCharImg);`
instead of 
`final BufferedImage image = ImageUtil.loadImageResource(getClass(), filePath);`
in JapanesePlugin.java, line 94
Is this alright?

ps:
I made these change becuase
-  I wanted to use all colors of japanese characters (instead of half, by using yellow instead of white, etc)
- no need to make PRs for when I update translations, which is going to happen very often

**Comment by YS-jack** [comment]
Would it help if I remove the deepL dependency?
If checking the deepL stuff is taking time, I thought removing it might be better for both of us, because ultimately I aim to translate everything by hand, and when thats achieved DeepL would be unnecessary. (although deepL will help users understand untill that happens)
I know reviews take time and everyone involved are volunteers, so I appreciate you guys works.

**Comment by YS-jack** [comment]
i would like to hault for some time
I foud "runelingual" plugin, which is still in development, but does pretty much the same thing but supports multiple languages in 1 plugin
it would be better if I collaborated with them, right?
it would take some time to migrate things, but if its for the better i will

**Comment by raiyni** [comment]
That would  be preferred, yes

**Comment by YS-jack** [comment]
hard choice but I think I should do that, more work now but less in the future
can i close this pr?

**Comment by raiyni** [comment]
got it, ty

**Pull request closed without merging**